Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,136 @@
+
+Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),
+because that requires managing an extra component +CRDs on top of the Prometheus operator.
+Moreover, this project doesn't have a lot of community support, and it's reached by small amount of users.
+Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.
+
+Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users from managing a standalone component within the stack.
+
+## Pitfalls of the current solution
+
+Using Alertmanager API Directly comes with drawbacks:
+
+* Teams have to build an automation to add silences in a centralized manner
+* There is no input validation in CI, which can lead to an invalid silence configuration
+* Teams need to manage expiration on their own
+* If a silence was deleted by accident, it will be permanent.
+
+# Goals
+
+* Provide a way for users to manage Alertmanager silences with a Custom Resource Definition (CRD).
+* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
+* Multi tenancy: Silences CR can be using kubernetes rbac.
+* Moving the Giant Swarm silence operator to prometheus operator would provide better community support and less fragmentation.
+* Improve user experience when managing silences, so users can use the same tools to manage their monitoring pipeline (kyverno, cr validations).
+
+## Audience
+
+* Users who use the Prometheus operator to offer monitoring-as-a-service to developers.
+* Users who want to manage silences the same way as for services running within the Kubernetes cluster.
+* Users who want a supported GitOps way of silences outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal do not cover automatic cleanup of expired Silence CRs
+* Refactoring of the other CRDs is not in scope for the first version
+* Adding silences to a remote Alertmanager cluster
+
+# How
+
+Creating a new namespaced Silence CRD that will act as an interface by adding silences via the Alertmanager API.
+
+A typical `Silence` resource could look like the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: Silence
+metadata:
+  name: my-silence
+  namespace: monitoring
+  labels:
+    test: value
+spec:
+  expiresAt: "2023-06-01 10:00"
+  matchers:
+    - matchType: "="
+      name: namespace
+      value: test-ns
+```
+
+The above resource will result in creating a silence in the Alertmanager that has the appropriate selector.
+
```

Comment:
What happens if a Silence CR cannot create a silence in AM? Can you describe certain error classes (e.g. invalid configuration, AM unavailable, etc.) and how they will be handled?

Reply:
Yes, the status sub resource, will has the error messages, 
Again, this is will be defined in the design document